src/pathfinder/yapf/yapf_ship_regions.cpp

```diff
@@ -0,0 +1,330 @@
+					pf.AddOrigin(GetWaterRegionPatchInfo(tile));
+				}
+			}
+		} else {
+			TileIndex tile = v->dest_tile;
+			pf.AddOrigin(GetWaterRegionPatchInfo(tile));
+		}
+
+		/* If origin and destination are the same we simply return that water patch */
+		std::vector<WaterRegionPatchDesc> path = { start_water_region_patch };
+		path.reserve(max_returned_path_length);
+		if (pf.HasOrigin(start_water_region_patch)) return path;
+
+		/* Find best path */
+		if (!pf.FindPath(v)) return {}; // Path not found
+
+		Node *node = pf.GetBestNode();
+		for (int i = 0; i < max_returned_path_length - 1; ++i) {
+			if (node != nullptr) {
+				node = node->m_parent;
+				if (node != nullptr) path.push_back(node->m_key.m_water_region_patch);
+			}
+		}
+
+		/* Draw the rest of the path black */
+		while (node != nullptr)
+		{
+			node = node->m_parent;
+			if (node != nullptr) {
+				auto patch = node->m_key.m_water_region_patch;
+				DEBUG_MarkRegion(patch.x, patch.y, patch.label, 3, true); // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+			}
+		}
+
+		assert(!path.empty());
+		DEBUG_MarkRegion(path.back().x, path.back().y, path.back().label, 5, true); // 0 = normal, 1 = red, 2 = blue, 3 = black, 4 = transparent
+		return path;
+	}
+};
+
+/** Cost Provider of YAPF for water regions */
+template <class Types>
+class CYapfCostRegionT
+{
+public:
+	typedef typename Types::Tpf Tpf;              ///< the pathfinder class (derived from THIS class)
+	typedef typename Types::TrackFollower TrackFollower;
+	typedef typename Types::NodeList::Titem Node; ///< this will be our node type
+	typedef typename Node::Key Key;               ///< key to hash tables
+
+protected:
+	/// to access inherited path finder
+	Tpf& Yapf() { return *static_cast<Tpf*>(this); }
+
+public:
+	/** Called by YAPF to calculate the cost from the origin to the given node.
+	 *  Calculates only the cost of given node, adds it to the parent node cost
+	 *  and stores the result into Node::m_cost member */
+	inline bool PfCalcCost(Node &n, const TrackFollower *tf)
+	{
+		n.m_cost = n.m_parent->m_cost + ManhattanDistance(n.m_key, n.m_parent->m_key);
+
+		/* Incentivise zigzagging by adding a slight penalty when the search continues in the same direction. */
+		Node *grandparent = n.m_parent->m_parent;
+		if (grandparent != nullptr) {
+			const DiagDirDiff dir_diff = DiagDirDifference(n.m_parent->GetDiagDirFromParent(), n.GetDiagDirFromParent());
+			if (dir_diff != DIAGDIRDIFF_90LEFT && dir_diff != DIAGDIRDIFF_90RIGHT) n.m_cost += 1;
+		}
+
+		return true;
+	}
+};
+
+/* We don't need a follower but the YAPF requires one */
+struct DummyFollower : public CFollowTrackWater {};
+
+/** Config struct of YAPF for route planning.
+ *  Defines all 6 base YAPF modules as classes providing services for CYapfBaseT.
+ */
+template <class Tpf_, class Tnode_list>
+struct CYapfRegion_TypesT
+{
+	/** Types - shortcut for this struct type */
```

Comment:
Bit odd comments. Repeating of the variable name isn't needed, but seems they can use a freshening up in general :)

Reply:
I've improved the style of this struct for the yapf_ship and yapf_ship_regions classes.